### PR TITLE
Substitute cppflow submodule with conan package

### LINKED ISSUE #342: Add tensorflow API installation instructions to readme
Needed only if we run with the ML models

### LINKED ISSUE #352: Add tensorflow install path to environment in macos CI workflow
Adding it into the environment in https://github.com/UCL/install-tensorflow-action/pull/2 and https://github.com/UCL/install-tensorflow-action/pull/3 seems to not have worked. Maybe it is in the wrong scope for the sopt CI. Try adding it manually in a separate step and see if that resolves the failing apple-clang workflow.

### PR DESCRIPTION
**Pull request by tkoskela**
In this PR we assume a cppflow conan package using https://github.com/UCL/cppflow/pull/8 exists and can be used to build sopt

### CODE REVIEW

**Comment by SJaffa** [comment]
Trying to match https://www.tensorflow.org/install/lang_c#linker

**Comment by SJaffa** [comment]
OK I have tried:

- Putting `echo "LD_LIBRARY_PATH=/usr/local/lib" >> $GITHUB_ENV` in it's own step after installing TF API
- Putting it in the cppflow package creation step
- Changing it to `echo "DYLD_LIBRARY_PATH=/usr/local/lib" >> $GITHUB_ENV` to match the [TensorFlow docs](https://www.tensorflow.org/install/lang_c#linker)
- Running this in a tmate session and manually setting all those environment variables

And I still get errors that `cppflow` can't find the tensorflow header files.

I created the little [example program](https://www.tensorflow.org/install/lang_c#example_program) from the tensorflow docs which should just check if the installation worked ok, and that did work! It found the `tensorflow/c/c_api.h` file which is one of the ones that `cppflow` specifically complained about not finding!!

So now I'm confused.

File: .github/workflows/cmake.yml [hunk 1]
```diff
@@ -132,6 +132,26 @@ jobs:
...
+      run: |
+        conan create ./cppflow/ -pr:h=default -pr:b=default -s compiler.libcxx=libstdc++11
+
+    - name: Create cppflow package on apple-clang
+      if: ${{ contains(matrix.cxx, 'clang++') }}
+      run: |
+        echo "LIBRARY_PATH=$LIBRARY_PATH:/usr/local/lib" >> "$GITHUB_ENV"
+        echo "DYLD_LIBRARY_PATH=/usr/local/lib" >> "$GITHUB_ENV"
+        cat $GITHUB_ENV
+        conan create ./cppflow/ -pr:h=default -pr:b=default
```
Comment (dstansby) [inline .github/workflows/cmake.yml:147]:
```suggestion
        echo $LIBRARY_PATH
        echo $DYLD_LIBRARY_PATH
        conan create ./cppflow/ -pr:h=default -pr:b=default
```

Comment (dstansby) [inline .github/workflows/cmake.yml:153]:
```suggestion
        echo $LIBRARY_PATH
        echo $DYLD_LIBRARY_PATH
        conan create ./cppflow/ -pr:h=default -pr:b=default
```

File: .github/workflows/cmake.yml [hunk 2]
```diff
@@ -81,12 +76,14 @@ jobs:
     - name: Install Dependencies on MacOS
       if: ${{ contains(matrix.os, 'macos') }}
       run: |
-        # Skip brew update until https://github.com/actions/setup-python/issues/577 is fixed
-        # Seems to be working, alghough the issue is still open
-        brew update
```
Comment (SJaffa) [inline .github/workflows/cmake.yml:86]: Do we want to keep the `brew update` step? That issue is still open but I don't think it is a problem for us at the moment.

File: .github/workflows/macdebug.yml [hunk 3]
```diff
@@ -28,21 +28,60 @@ jobs:
...
+        version: 1.59.0
+    
+    - name: Checkout cppflow repo
+      uses: actions/checkout@v3
+      with:
+        repository: UCL/cppflow.git
+        path: cppflow
+        ref: master
+
+    - name: Add Tensorflow paths to env
```
Comment (SJaffa) [inline .github/workflows/macdebug.yml]: This step was me trying to fix the issue that is now solved in `cppflow` so should probably remove this.

File: .github/workflows/cmake.yml (hunk 2 above)
Comment (tkoskela) [reply .github/workflows/cmake.yml:86]: I would remove it for now. Something has changed on GitHub actions and the brew update was spending a lot of time updating unrelated packages. Here is an [example](https://github.com/astro-informatics/sopt/actions/runs/5133162235/jobs/9235386890) that took 12 minutes to run the brew install step. Running on the package versions that come with the macos runner seems fine. If something breaks we will have to update packages one by one.

Withtout brew update [it takes 40 seconds](https://github.com/astro-informatics/sopt/actions/runs/5142619835/jobs/9256578473) to brew install the packages we need.

**Comment by SJaffa** [comment]
I've updated the installation instructions to use the `cppflow` conan package, are there any other updates we need to do in the readme to match the way this branch builds?

**Comment by tkoskela** [comment]
I added the `cppflow` option to the list of CMake options. There's also some outdated stuff in the `Dependencies Installation` section.

**Comment by SJaffa** [comment]
Is this one ready to merge now or are we still working on bits?

**Comment by tkoskela** [comment]
It's ready, if someone can review and possibly approve that would be great

**Review by SJaffa — APPROVED** [review]

**Review by mmcleod89 — APPROVED** [review]

**Pull request merged**